DO NOT MERGE  Zoltan2icesheet - Add connectivity checking functionality for ice sheet simulations

ibogle opened the pull request:
<!---
Be sure to select `develop` as the `base` branch against which to create this
pull request.  Only pull requests against `develop` will undergo Trilinos'
automated testing.  Pull requests against `master` will be ignored.

Provide a general summary of your changes in the Title above.  If this pull
request pertains to a particular package in Trilinos, it's worthwhile to start
the title with "PackageName:  ".

Note that anything between these delimiters is a comment that will not appear
in the pull request description once created. Most areas in this message are
commented out and can be easily added by removing the comment delimiters.

Please make sure to mark:
* Reviewers
* Assignees
* Labels

Replace <teamName> below with the appropriate Trilinos package/team name.
-->
@trilinos/zoltan2 

## Motivation
<!--- 
Why is this change required?  What problem does it solve? Please link to a github 
issue that describes the problem/issue/bug this PR solves.
-->
This change introduces functionality that aides the convergence of land-ice ice-sheet simulations.
<!---
If applicable, let us know how this merge request is related to any other open
issues or pull requests:

## Related Issues

* Closes 
* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of 
-->


## Stakeholder Feedback
<!--- 
If a github issue includes feedback from the relevant stakeholder(s), please link it.  
If the stakeholder(s) communicated that feedback through a different medium, please note that you did so.
-->

## Testing
<!---
Please confirm that any classes or functions in the Trilinos library that this PR touches are 
exercised by at least one test in Trilinos.  Please specify which test that is.  For untestable 
changes (e.g. changes to the nightly testing system) or changes to Trilinos tests, please say "N/A".

-->
Testing is done through an executable in the zoltan2/test/connectivity/ directory. Tests include real inputs, small adversarial inputs, and a test to ensure the code can handle distributions where one process gets no vertices. There are additional tests that can be run if Trilinos was configured with ParMETIS that ensure that the code can handle vertex distributions that are similar to our expected use cases. Tests have been verified with int and long long as the global ordinal type.
<!--- 
## Additional Information
Anything else we need to know in evaluating this merge request?
 -->

mhoemmen commented:
GitHub doesn't let us review the changes -- it says the diff is too big to generate.  Would it make sense to rebase or cherry-pick the changes against current develop, in order to help GitHub?

kddevin commented on packages/zoltan2/test/connectivity/iceSheetIntegration.cpp:
```diff
@@ -0,0 +1,227 @@
+  int numLocalBoundaryEdges = 0;
+  for(int i = 0; i < num_global_boundary_edges; i+=2){
+    for(int j = 0; j < nlocal; j++){
+      if(boundary_edges_global[i] == vtxIDs[j] || boundary_edges_global[i+1] == vtxIDs[j]){
```
Rather than linear search, how about using Map's check for global IDs?  Map->getLocalElement(boundary_edges_global[i]) != Teuchos::OrdinalTraits<LocalOrdinal>::invalid() ==> processor has the vtx
https://docs.trilinos.org/dev/packages/tpetra/doc/html/classTpetra_1_1Map.html#a11c6a4585d616718eeb5c16d8a492511

ibogle replied:
I'll use crsgraph->getRowMap() as the map in both cases.

ibogle replied:
I'll change this to use crsgraph->getRowMap() as the map for the ownership check.

kddevin commented on packages/zoltan2/test/connectivity/iceSheetIntegration.cpp:
```diff
@@ -0,0 +1,227 @@
+  int edgecounter = 0;
+  for(int i = 0; i < num_global_boundary_edges; i+=2){
+    for(int j = 0; j < nlocal; j++){
+      if(boundary_edges_global[i] == vtxIDs[j] || boundary_edges_global[i+1] == vtxIDs[j]){
```
Ditto

kddevin commented on packages/zoltan2/test/connectivity/Zoltan2_IceUtil.h:
```diff
@@ -0,0 +1,187 @@
+#include<sstream>
+#include<string>
+
+#include "graph.h"
```
I thought the name of this include file changed.  Would an application need to include it?

ibogle replied:
This include was overlooked after I changed the name of the file. It will be removed.

kddevin commented on packages/zoltan2/test/connectivity/iceSheetIntegration.cpp:
```diff
@@ -0,0 +1,227 @@
+  //need to read in problem specific files here, start out with zeroed out arrays
+  int nlocal = (int) inputGraphAdapter->getLocalNumVertices();
+  bool* basalFriction = new bool[nlocal];
+  const gno_t* vtxIDs = new gno_t[nlocal];
```
1.  When getting a view, you do not pre-allocate the memory.  A pointer is returned in vtxIDs.  This allocated memory will be leaked.
2.  Why not get the vtxID info from the CrsGraph's Map?  Or just use the Map's methods to accomplish what you need, instead of using vtxIDs?

ibogle replied:
2. is a very good point, I will switch to using the CrsGraph's RowMap to do the conversion from local to global identifiers and get rid of the vtxIDs array.

kddevin commented on packages/zoltan2/test/connectivity/iceSheetIntegration.cpp:
```diff
@@ -0,0 +1,227 @@
+  std::cout<<me<<": global_boundary_edges = "<<num_global_boundary_edges<<" localBoundaryEdges = "<<2*numLocalBoundaryEdges<<"\n";
+  gno_t* boundaryEdges = new gno_t[2*numLocalBoundaryEdges];
+  
+  for(int i = 0; i < 2*numLocalBoundaryEdges; i++){
```
Why initialize if going to overwrite?

ibogle replied:
I think these may have been left over sanity checks from debugging. I will remove them.

kddevin commented on packages/zoltan2/test/connectivity/iceSheetIntegration.cpp:
```diff
@@ -0,0 +1,227 @@
+    boundaryEdges[i] = 0;
+  }
+
+  for(int i = 0; i < nlocal; i++){
```
Ditto.

kddevin commented on packages/zoltan2/test/connectivity/meshData/README.txt:
```diff
@@ -0,0 +1,42 @@
+
+*-basal-friction.ascii
+line 1:  #meshvertices
+remaining lines:  1 if vertex is grounded; 0 otherwise
```
ant*km-basal-friction.ascii contains floating point numbers.
Are the files wrong?  Or is this documentation wrong?

ibogle replied:
The documentation is correct for files created by hand. I will add a note that real inputs have floating point numbers for the basal friction.

kddevin commented on packages/zoltan2/test/connectivity/meshData/README.txt:
```diff
@@ -0,0 +1,42 @@
+
+File formats:
+
+*.quad.msh -- mesh files with quad elements
```
ant*km-borders.quad.msh is a member of *.quad.msh, but has different format (as shown below)

kddevin replied:
Also, there are no other *.quad.mesh files besides *-borders.quad.msh

ibogle replied:
This documentation was written before we made the switch to MatrixMarket files, so this part is outdated.

kddevin commented on packages/zoltan2/test/connectivity/meshData/README.txt:
```diff
@@ -0,0 +1,42 @@
+   for each vertex v flagged as grounded in *-basal-friction.ascii, 
+   if any element incident to v has at least s grounded vertices, v is grounded.
+
+NOTE:  Currently, these files are zero-based; it would be better for them to 
```
Agreed.  Why not convert them?

ibogle replied:
I have converted the tests and the validation to account for 1-indexed vertices.

kddevin commented on packages/zoltan2/test/connectivity/meshData/README.txt:
```diff
@@ -0,0 +1,42 @@
+NOTE:  Currently, these files are zero-based; it would be better for them to 
+match the mesh files (one-based) so that we could more easily hand-check them.
+
+***NOTE***:  
```
Is this still TODO?

ibogle replied:
No. My validation checks false positives and false negatives.

kddevin commented on packages/zoltan2/test/connectivity/meshData/README.txt:
```diff
@@ -0,0 +1,42 @@
+should be removed by distributed does not flag it
+Need to check removing same and keeping same.  (See below.)
+
+***NOTE***:  
```
Is this still TODO?

ibogle replied:
I don't think so. I'll double check the answers for that, but I'm pretty sure the test is passing correctly.

ibogle replied:
The answers file for that test is empty, so that is no longer TODO.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,301 @@
+      typedef typename Adapter::userCoord_t userCoord_t;
+      typedef Tpetra::Map<lno_t, gno_t> map_t;
+      
+      //Arguments: problemComm is the communicator we will use for the propagation
```
Since this is the main interface for user, it needs better and doxygenized documentation of inputs.

ibogle replied:
I think this has been sufficiently addressed. The wording may need some clarification.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,301 @@
+	buildModel(flags);
+      }
+   
+      //This function returns a number of flags consistent with the 
```
Since this is the main interface for user, it needs better and doxygenized documentation of outputs.  What is returned?  How does one interpret the flags' values?  Etc.

ibogle replied:
I have added better documentation.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,301 @@
+      const RCP<const base_adapter_t> adapter;
+      void buildModel(modelFlag_t &flags);
+
+      bool* grounding_flags;
```
I'm concerned about storing user's pointers here.  What if user changes them between constructor and getDegenerateFeatureFlags?
Perhaps this capability should be a utility function, rather than a class. 
We can discuss.

ibogle replied:
That is a good point, it shouldn't be on the user to know which pointers they cannot touch between constructing the object and calling the propagation function. I think it makes more sense as a templated utility function; it would be easier and more straightforward to use that way. It should be fairly straightforward to change it, but I will wait until I hear back from you to change this.

kddevin replied:
Two possible approaches:

1. The constructor computes the solution and, thus, avoids storing pointers to the application's data.  It stores the result in an array RCP, and then returns its raw pointer in getDegenerateFeatureFlagsView.  This is analogous to Zoltan_PartitioningProblem::getPartListView.

2.  Instead of IceProp being a class, it could be a utility looking something like this:
      DetectDegenerateVertices(const RCP<const Comm<int> > &problemComm__,
	      const RCP<const GraphAdapter<user_t,userCoord_t> > &adapter__, 
	      bool* basalFriction, gno_t* boundary_edges__, int num_boundary_edges__, gno_t *degenerateFlags)
where the user is expected to allocate the result array to the correct size for num_local_vertices
(Use const to indicate which arguments are not changed by the utility.)

ibogle replied:
I think implementing the first approach would be less development effort. Just to make sure I'm clear, when you say return its raw pointer, you mean I would call getRawPtr() on the ArrayRCP?

ibogle commented on packages/zoltan2/test/connectivity/iceSheetIntegration.cpp:
```diff
@@ -0,0 +1,227 @@
+  int* grounded_flags_global;
+  gno_t* boundary_edges_global;
+
+  int nglobal = 0;
```
I'll be changing all of these instances to size_t.

ibogle commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,318 @@
+      *          -1 for a vertex that has no connection to the ground (e.g. an iceberg).
+      *          <vtxID> for a vertex that is in a floating hinge, where vtxID is the hinge vertex.
+      */
+      int* getDegenerateFeatureFlags();
```
I think this return type should change. We make use of negative integers, so is it safe to use the global ordinal type?

kddevin replied:
Given the documentation, you are correct that int is not appropriate because you are returning global ordinals in some cases.
Global ordinal type can also be a challenge; it is almost always signed, but Trilinos does offer Tpetra_INST_INT_UNSIGNED_LONG_LONG, which would have unsigned global ordinals.

I'm trying to think of something clever, assuming we return data as gno_t.  One option:
define constants Zoltan2::AttachedToGround = std::numeric_limits<gno_t>::max and Zoltan2::Floating = std::numeric_limits<gno_t>-1  and return those instead of -2 and -1?  That's not perfect, as it prevents the user from using these two global IDs (but we can check for that condition in the Map).

What clever ideas do you have?

ibogle replied:
Your suggestion sounds like it would be the best way to implement the current approach with unsigned numbers in mind. With absolute scalability in mind, it seems like this approach would be problematic for cases where the number of global IDs used approaches the limit of the global ordinal type. Now, that is very unlikely to happen, but I'm not sure how many ways we'd be able to resolve that issue without sacrificing some functionality. It seems like our flags would be ambiguous to the user in that case, unless I'm misunderstanding your note about using the map to check for the global IDs, which is entirely possible.

Maybe it would be best to return two arrays, as we are currently attempting to convey two types of information with the same array. For instance, an integer array to indicate a vertex's status, where 0 = keep, 1 = floating, and 2 = hinged, and then a separate global ordinal array that indicates the hinge vertex where applicable. This is more memory-intensive, and since very few vertices are typically removed it may not be as efficient as our previous approach, but it allows all global IDs to be used, and still conveys the same information. This may also result in more readable code, as it would be clear that the different arrays indicate different things (we aren't mixing sentinel values and vertex identifiers). I'm thinking in this approach we would use just some default value (e.g. 0) for a vertex that does not have a status of 2, and <vtxID> for a vertex with status 2. One drawback is that it would be easier to misinterpret these hinge identifiers, as all of the entries are valid IDs.

kddevin replied:
OK, if you are going to return two arrays, please use the utility interface in which the user is responsible for allocating the arrays.  It might be good for them to be Teuchos::ArrayView, then, so that we can error-check that they have sufficient size.  Similarly, we can use ArrayView for the input arguments (basal friction, boundary edges).

ibogle replied:
Noted. I will get to work on this today.

ibogle replied:
I changed the return type of the propagation function so that it returns the ArrayRCP of vertex labels. I am then going to interpret the labels in the utility function to fill in the two output ArrayViews.

kddevin replied:
I don't understand. The ArrayRCP was needed if we used a class interface with getDegenerateFeatureFlags.  But with a utility interface, we aren't saving pointers to any data, so we don't need an ArrayRCP, right?

ibogle replied:
The class that does the actual propagation stores the labels in an ArrayRCP<const VtxLabel>, so the propagation function currently returns that array of labels to the utility function. Then, the utility function will interpret the labels and set the output ArrayViews appropriately.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.hpp:
```diff
@@ -0,0 +1,713 @@
+          public:
+            size_t n;
+            size_t m;
+            lno_t* out_array;
```
All fields of graph data structure need comments.  What are they?

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.hpp:
```diff
@@ -0,0 +1,713 @@
+	// Struct representing a vertex label.
+	// We define our own "addition" for these labels.
+	// Later, we'll create a Tpetra::FEMultiVector of these labels.
+	class IcePropVtxLabel {
```
All fields of data structure need comments.  What are they?

kddevin replied:
I was surprised that the graph is templated on lno_t, but the vertex labels are not.
Aren't the senders and labels actually vertex IDs?  In fact, I thought they were global vertex IDs.  If so, they should be defined with the templated type gno_t.

ibogle replied:
This was a similar issue to the return type of the flags. I use a sentinel value to indicate an empty label, this class should be templated on gno_t. In the spirit of the other solution, a way around the sentinel value could be to add two fields indicating whether or not a label has a valid value in it.

ibogle replied:
I use -1 to represent unused currently. The labels contain global identifiers, so they should be templated on the global ordinal type.

ibogle replied:
I'm currently reworking these last changes, I have templated the label class on both gno_t and lno_t, because the id field is actually a local identifier.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.hpp:
```diff
@@ -244,8 +281,8 @@ namespace Kokkos {
 template<typename Ordinal>
-struct SerializationTraits<Ordinal, Zoltan2::IcePropVtxLabel> :
-       public Teuchos::DirectSerializationTraits<Ordinal, Zoltan2::IcePropVtxLabel>
+struct SerializationTraits<Ordinal, Zoltan2::IcePropVtxLabel<Tpetra::Map<>::global_ordinal_type, Tpetra::Map<>::local_ordinal_type> >:
```
Everything else in Trilinos has local ordinal first and then global ordinal.  Switching the order for one algorithm is bound to cause trouble someday.

ibogle replied:
I will switch that!

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.hpp:
```diff
@@ -398,7 +436,7 @@ class iceSheetPropagation {
         lno_t* outs = g->out_vertices(curr_node.id);
         for(int i = 0; i < out_degree; i++){
-	  IcePropVtxLabel neighbor = femvData[outs[i]];
+	  IcePropVtxLabel<gno_t,lno_t> neighbor = femvData[outs[i]];
```
Maybe define a type so that you don't have to type the template arguments everywhere.

ibogle replied:
Will do.

ibogle commented:
I think the switch to a utility function interface got rid of all the -Wshadow warnings, I compiled with that flag, but I didn't see any relevant warnings. I think I have covered all the issues you have raised so far. I updated the documentation for the utility function, removed the old class definition, and made sure all the tests pass, including the ParMETIS tests. I added comments, and fixed the order of IcePropVtxLabel's template arguments.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+      if(i < nVtx){
+        grounding[i] = basalFriction[i];
+      }
+      else grounding[i] = false;
```
Is it necessary for grounding to store false values for the ghosts?  
If we don't need the grounding info for the ghosts, we could avoid the entire copy from basalFriction to grounding.
See above note in Zoltan2_IceProp.hpp.

ibogle replied:
I do not think this is necessary. I have reduced the grounding size to just the locally owned vertices.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+    femv->beginFill();
+    //set the member variable that stores femv->getData(0);
+    femvData = femv->getData(0);
+    for(lno_t i = 0; i < nLocalOwned + nLocalCopy; i++){
```
Why does this for-loop loop over the ghost entries (nLocalCopy)?
The boundary flag test (line 51) assures that only owned vertices are marked as articulation points.
The grounding flags test (line 54) will be false for all ghosts (see comment in Zoltan2_IceSheet.hpp).
So nothing will be done for ghost vertices here, right?
Should this loop go over only owned vertices?
Or should grounding_flag info in Zoltan2_IceSheet.hpp be updated with ghosts' grounding info?

ibogle replied:
This loop initializes the label datastructure for all vertices, so I do need to loop through ghost entries. However, I do not need to use the grounding information for ghost entries, just set the global ID for future reference. I have changed this as well so the smaller grounding array will not cause issues here. Tests pass.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+ 
+    //communicate boundary edges back to owning processors
+    gno_t num_boundary_edges = boundary_edges.size()/2;
+    std::vector<int> owners_vec;
```
Why are the copies owners_vec and boundary_vec needed?
Why not use boundary_edges directly instead of making boundary_vec and boundary?
Why not declare Teuchos::Array owners(boundary_edges.size(), 0)?

ibogle replied:
I wasn't aware of a better way to do this, and there are likely many places that I have clumsy usage of Teuchos datastructures. I will clean this up as you suggest.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+    int nprocs = problemComm->getSize();
+  
+    //send boundary edges to any remote processes that own an endpoint
+    std::vector<int> sendcnts;
```
Why make sendcnts and recvcnts be std::vector, followed by creation of ArrayView instead of making them Teuchos::Array?  
Teuchos::Array sendCount(nprocs, 0);
Teuchos::Array recvCount(nprocs, 0);
Then you get the ArrayView as sendCount()

ibogle replied:
This has been done in each instance.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+    Teuchos::ArrayView<int> recvCount = Teuchos::arrayViewFromVector(recvcnts);
+  
+    //create sdispls and sentcount to build sendbuf
+    std::vector<int> sdispls;
```
Same here.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+ 
+    //Do the final communication back to each remote process 
+    Teuchos::ArrayRCP<gno_t> recvbuf;
+    Teuchos::ArrayView<const gno_t> sendBuf = Teuchos::arrayViewFromVector(sendbuf);
```
Same here.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+    //hash for the boundary-edge set
+    struct pair_hash {
+      inline std::size_t operator()(const std::pair<gno_t,gno_t>& v) const {
+        return v.first * 10 + v.second;
```
Why this choice of hash function?

ibogle replied:
I didn't spend too much time on this hash function. I think my reasoning behind it was that an edge from A to B should be unlikely to get the same hash as an edge from B to A. Although, I do insert the edges so that they are in sorted order, so that case should never occur. I am very much open to changing this.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+   *  \param (input) basalFriction An ArrayView of boolean flags for each local vertex with true representing a grounded vertex and false representing a floating vertex.
+   *                        These indicies must be consistent with the local vertex identifiers for the input graph.
+   *  \param (input) boundary_edges An ArrayView of global vertex identifiers representing all of the edges that the current process knows about which are on the boundary between the ice and the water. 
+   *                        The IDs must be present in the overlap map of the current process if the ID is not owned by this process. Edges should not be specified more than once.
```
Does "IDs must be present in the overlap map of the current process" mean that a process can handle a boundary edge between a ghost vertex and a vertex on another process?  Or must at least one vertex of each edge be owned by the process?

Regarding "Edges should not be specified more than once":  Say proc 0 owns ID A, and proc 1 owns ID B.  Does proc 0 give edge A->B and proc 1 give edge B->A?  Or is that an error?

ibogle replied:
"IDs must be present in the overlap map of the current process" means that if there is an ID in the boundary edges, the current process must know which process owns that ID. This constraint was made from my understanding of Mauro's explanation of which boundary edges are easy to provide, so this should be what Mauro is expecting. If there is an edge from a ghost vertex to an unknown vertex (a vertex whose owner is unknown), the current process may not be able to easily send the edge to both of the processes that need it.

"Edges should not be specified more than once" is purely a local constraint. Globally, edges may be specified more than once, but I use a set to ensure that remote boundary edges are not counted more than once, as they could exist on more than one process. I can add this to the comment to make it more clear.

ibogle replied:
So to answer your second question, A->B on proc 0 and B->A on proc 1 is fine.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+    //create the set to see if we've counted this boundary edge before
+    std::unordered_set<std::pair<gno_t,gno_t>,pair_hash> edge_set;
+
+    int* local_boundary_counts = new int[nVtx];
```
The rest of the code uses std::vector or Teuchos::Array; why does this array use C++ array?

ibogle replied:
I have changed it to a Teuchos::Array.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+      if(gs == ICEPROPGS_FULL) vertex_status[i] = -2;
+      else if(gs == ICEPROPGS_HALF){
+        vertex_status[i] =0;
+        hinge_vertices[i] = curr_node.first_label;    
```
Don't we need to set hinge_vertices in all cases?  Seems dangerous to return it uninitialized for other cases.

ibogle replied:
hinge_vertices only has meaning if vertex_status is 0. If the vertex status is not zero, we cannot detect a hinge vertex after propagation is complete. The comments for the output arguments explain the usage of vertex_status and hinge_vertices, but I agree it is not ideal, as hinge_vertices will always have a valid-looking identifier in each entry.

Previously I had returned a signed integer, with negative values representing vertex_status, and non-negative values representing hinge vertex IDs, but this approach does not work with unsigned return types. This new approach solves that issue, but makes the output arguments a little less user-friendly. If you have any ideas on how to improve this, please let me know.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+      std::cout<<"Vertex "<<gid<<", First label="<<curr_node.first_label<<", first sender="<<curr_node.first_sender<<", first used="<<curr_node.first_used;
+      std::cout<<", second label="<<curr_node.second_label<<", second sender="<<curr_node.second_sender<<", second_used="<<curr_node.second_used<<"\n";
+      IcePropGrounding_Status gs = curr_node.getGroundingStatus();
+      if(gs == ICEPROPGS_FULL) vertex_status[i] = -2;
```
Instead of -2, 0, -1 for status, can we define an enumerated type against which the user can test?  Something like Zoltan2::IceGrounded, Zoltan2::IceFloating, Zoltan2::IceHinged?

ibogle replied:
I have created this enumerated type.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+    for(size_t i = 0; i < nEdge; i++){
+      out_edges_lid[i] = mapWithCopies->getLocalElement(out_edges[i]);
+    }
+    icePropGraph<typename map_t::local_ordinal_type>* g = new icePropGraph<typename map_t::local_ordinal_type>({nVtx, nEdge, out_edges_lid,out_offsets});
```
Why should g be a pointer instead of an object?

ibogle replied:
I have changed the pointer to be an object.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+    for(size_t i = 0; i < g->n; i++){
+      IcePropVtxLabel<lno_t,gno_t> curr_node = labels[i];
+      gno_t gid = mapWithCopies->getGlobalElement(curr_node.id);
+      std::cout<<"Vertex "<<gid<<", First label="<<curr_node.first_label<<", first sender="<<curr_node.first_sender<<", first used="<<curr_node.first_used;
```
I/O like this needs to be removed or guarded with a debug level as in Zoltan2_AlgPuLP.hpp

ibogle replied:
It has been removed.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+
+    //convert adjacency array to use local identifiers instead of global.
+  
+    typename map_t::local_ordinal_type* out_edges_lid = new typename map_t::local_ordinal_type[nEdge];
```
The rest of the code uses std::vector or Teuchos::Array; why does this array use C++ array?

ibogle replied:
It was due to the graph class having a C++ array as a member. I've fixed that, so the graph data structure uses Teuchos::Array, so this line has also been changed to Teuchos::Array.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp:
```diff
@@ -0,0 +1,306 @@
+    for(size_t i = 0; i < nEdge; i++){
+      out_edges_lid[i] = mapWithCopies->getLocalElement(out_edges[i]);
+    }
+    icePropGraph<typename map_t::local_ordinal_type>* g = new icePropGraph<typename map_t::local_ordinal_type>({nVtx, nEdge, out_edges_lid,out_offsets});
```
I dislike one-character variable names except for loop indices.  It is difficult to search for them when working through files.  Please change g, n, m to have meaningful names.  Thanks!

ibogle replied:
I have gone through and changed all instances of g to local_graph, n to nVtx, and m to nEdge. The code compiles and tests run successfully, so I'm fairly sure I got them all.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+      neighbor.second_sender = curr_node_gid;
+      neighbor.second_used = curr_node.second_used;
+      neighbor.bcc_name = curr_node.bcc_name;
+      return;
```
Does the updated neighbor need to be added to a queue (as at line 266)?
If not, please document why.

ibogle replied:
I am not sure why those returns were there, but I think they should be added to the queue. I may have changed when the queue was updated at some point, and forgot to take the returns out. I have removed them, and the tests still pass.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+          }
+        }
+      }
+      return;
```
Does the updated neighbor need to be added to a queue (as at line 266)?
If not, please document why.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+  }
+  
+  //propagation rules to find biconnected components 
+  Teuchos::ArrayRCP<const scalar_t> bccPropagate(){
```
Should this function be here?  I do not see it exercised anywhere.  Is it ready for inclusion in Zoltan2?  Is it the code for the SIAM CSC talk?  If so, please add tests for it.  I will not carefully review this function now, but if you tell me it will stay, I will do a careful review.  Note the comment on switching below.

ibogle replied:
This is the function for finding Biconnected components. It is an implementation of the algorithm I presented at CSC. I would be alright with moving this to a different branch after merging this PR and going from there.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+  Teuchos::ArrayRCP<const scalar_t> femvData;
+};
+
+//std::queue<iceSheetPropagation<Tpetra::Map<>>::lno_t> iceSheetPropagation<Tpetra::Map<>> ::regQueue;
```
Delete these comments; the queues were moved.

ibogle replied:
Done.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.hpp:
```diff
@@ -0,0 +1,378 @@
+namespace Zoltan2{
+        
+        template<typename MAP>
+        class iceSheetPropagation;
```
Why is this forward declaration needed?  I don't see any references to iceSheetPropagation in this file.

ibogle replied:
I have removed this forward declaration.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.hpp:
```diff
@@ -0,0 +1,378 @@
+        class iceSheetPropagation;
+	//Graph datastructure that represents the local
+	//graph in ice sheet propagation problems
+        template<typename lno_t>
```
Please use consistent formatting in all files:  two-space indentation and no tabs.
I prefer 80-columns, too.

ibogle replied:
I'll go through and double check my formatting. I'll do my best to keep lines to 80 columns.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.hpp:
```diff
@@ -0,0 +1,378 @@
+	class icePropGraph {
+          public:
+            //n represents the number of vertices in the local graph
+            size_t n;
```
nVtx and nEdge would be clearer and more searchable than n and m.

ibogle replied:
I have made this change.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+#include "Tpetra_Core.hpp"
```
Add Trilinos header boilerplate.

ibogle replied:
I copied the header boilerplate from one of the coloring files.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+#define ZOLTAN2_ICEPROP_
+
+namespace Zoltan2{
+template<typename MAP>
```
Why is this class templated on Map instead of local and global ordinal?

ibogle replied:
Should I only template on local and global ordinal? Is it alright to fix the map type as a Tpetra Map?

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.hpp:
```diff
@@ -0,0 +1,378 @@
+          //(this is necessary for unsigned global types)
+          bool  second_used;
+
+          //this field may potentially be used to find biconnected components
```
Are we not sure whether bcc_name will be used to find BCCs?
Or are you saying that this field is needed to find BCCs?

ibogle replied:
This field is needed to find BCCs. I think this is an older comment, I can update it to be more accurate.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.hpp:
```diff
@@ -0,0 +1,378 @@
+                                    int second_ = -1, int second_sender_ = -1, bool second_used_ = false,
+                                    bool art_ = false, int bcc_name_ = -1) { 
+	    id = idx_;
+	    if(id == -1){
```
Is this debugging output?  Or an error condition?  Should it ever occur? 
The code should be able to run silently, so if this is debugging output, please guard it appropriately.  If it is an error, please specify the nature of the error in the output so that a user can address it.

ibogle replied:
This is something that should never occur. During normal running of the code, it should not be possible, so I think it is safe to remove the output. I have never seen that statement execute.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.hpp:
```diff
@@ -0,0 +1,378 @@
+            bcc_name = -1;
+	  }
+
+	  IcePropVtxLabel(volatile const IcePropVtxLabel& other){
```
Add comment about why all these variants of copy constructors and assignment overloads were needed.

ibogle replied:
Done.

kddevin commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+    if(curr_gs == ICEPROPGS_NONE) return;
+    //if the current node is full (and not an articulation point), pass both labels on
+    if(curr_gs == ICEPROPGS_FULL && !curr_node.is_art){
+      neighbor.first_label = curr_node.first_label;
```
Would using the overloaded += operator here provide greater readability?

ibogle replied:
The += operator considers fewer cases than the give_labels function. I'm not sure using the += operator would provide greater readability, especially since I've had issues before with the queues and cyclic references with the Vertex Label class. Making the += operator have the same functionality as give_labels may lead to more complexity that is distributed over a few files. i think having the propagation rules and functions all in one place makes the code easier to understand.

kddevin commented on packages/zoltan2/test/temp/vtxlabel.cpp:
```diff
@@ -0,0 +1,260 @@
+#include "Tpetra_Core.hpp"
```
You can delete this test.  femv.cpp does all the same and more (and has better comments).  Thanks.

ibogle replied:
Done.

kddevin commented on packages/zoltan2/test/connectivity/meshData/README.txt:
```diff
@@ -0,0 +1,35 @@
+
+*-borders.quad.msh
+line 1:  two meaningless words
+line 2:  #meshvertices (<== always zero) #elements (<== always zero) #borderEdges 
```
Why have always zero values?

kddevin commented on packages/zoltan2/test/connectivity/meshData/bigrectangle-borders.quad.msh:
```diff
@@ -0,0 +1,21 @@
+big rectangle
```
I like using this line as a description of the test.

ibogle replied:
After generating a few of these by hand, I made a boundary-edge generator, which is why many of the descriptions weren't very helpful.

kddevin commented on packages/zoltan2/test/connectivity/meshData/chaineddoublehinge-borders.quad.msh:
```diff
@@ -0,0 +1,31 @@
+big rectangle
```
Wrong description?

ibogle replied:
It has been updated.

kddevin commented on packages/zoltan2/test/connectivity/meshData/noedges-answers.txt:
```diff
@@ -0,0 +1,4 @@
+1
```
This file says all four vertices should be removed.  But the basil friction file says each point is grounded.  Grounded points should not be removed, right?  If the code removes them, that would be an error.

ibogle replied:
From my understanding, standalone vertices in the application are to be removed. Even though each vertex is grounded, there is no other connection to ground so each vertex should be removed.

ibogle replied:
In other words, being initially grounded does not guarantee that the vertex will be kept. Similarly, a mesh element with a single vertex that is initially grounded would have all vertices in that element removed.

kddevin commented on packages/zoltan2/test/connectivity/meshData/noedges-basal-friction.ascii:
```diff
@@ -0,0 +1,5 @@
+4
```
This file says all four vertices are grounded.  But the answers file says they should be removed.  Grounded points should not be removed, right? If the code removes them, that would be an error.
Perhaps this test should have some grounded vertices and some not-grounded vertices; the code should remove only the not-grounded vertices.

ibogle replied:
Vertices that are initially grounded may not necessarily be kept. In this case, they are removed because there are no edges, so it is impossible for them to be sufficiently grounded.

kddevin commented on packages/zoltan2/test/connectivity/meshData/test1-borders.quad.msh:
```diff
@@ -0,0 +1,10 @@
+nonsense line
```
Add a meaningful description

ibogle replied:
it has been updated.

kddevin commented on packages/zoltan2/test/connectivity/meshData/test1-borders.quad.msh:
```diff
@@ -0,0 +1,10 @@
+nonsense line
+0 0 8
```
This file does not follow the file format, which prescribes that line 3 must be blank.  Does it actually run correctly?  If so, why?

ibogle replied:
The format has changed, so this is no longer an issue.

kddevin commented on packages/zoltan2/test/connectivity/meshData/test1-diagram.ascii:
```diff
@@ -0,0 +1,5 @@
+    |   |
+7---4---3
+|   |
+6---5
```
Nice!

kddevin commented on packages/zoltan2/test/connectivity/meshData/three-borders.quad.msh:
```diff
@@ -0,0 +1,5 @@
+nonsense line
```
Add a meaningful description

ibogle replied:
Done.

kddevin commented on packages/zoltan2/test/connectivity/meshData/three-borders.quad.msh:
```diff
@@ -0,0 +1,5 @@
+nonsense line
+0 0 3
```
This file does not follow the file format, which prescribes that line 3 must be blank.  Does it actually run correctly?  If so, why?

ibogle replied:
The format has changed, so this is no longer an issue.

kddevin commented on packages/zoltan2/test/connectivity/meshData/tictactoe-borders.quad.msh:
```diff
@@ -0,0 +1,23 @@
+Nonsense line
```
Add a meaningful description

ibogle replied:
Done

kddevin commented on packages/zoltan2/test/connectivity/meshData/worstcase-borders.quad.msh:
```diff
@@ -0,0 +1,123 @@
+Nonsense line
```
Add a meaningful description

ibogle replied:
Done.

kddevin commented on packages/zoltan2/test/temp/CMakeLists.txt:
```diff
@@ -103,6 +103,21 @@ TRIBITS_ADD_EXECUTABLE_AND_TEST(
+    SOURCES femv.cpp
+    COMM mpi serial
+  )
+  TRIBITS_ADD_EXECUTABLE_AND_TEST(
```
OK to remove this test.

ibogle replied:
Done.

ibogle commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.cpp:
```diff
@@ -0,0 +1,94 @@
+
+#ifdef HAVE_TPETRA_EXPLICIT_INSTANTIATION
+
+#warning "KDDKDD HELLO"
```
Done.

ibogle commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+    femv = rcp(new femv_t(mapOwned, importer, nVec, true)); 
+    femv->beginFill();
+    //set the member variable that stores femv->getData(0);
+    femvData = femv->getData(0);
```
I will change this so that the femvData field gets updated with a call to femv->getData(0) after each switchActiveMultiVector and begin/end fill.

ibogle commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+          label.first_label = -1;
+          label.first_sender = -1;
+          label.bcc_name = -1;
+          femv->switchActiveMultiVector();
```
i need to take a look at my switching in general, I'm not sure I knew exactly what I was doing. I'll make sure only necessary switches are used, and that those are commented.

ibogle commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+      bfs_prop();     
+    }
+    //return the final labels for each owned vertex
+    return femvData;
```
I have altered the prop function to take vertex_status and hinge_vertices as input arguments, and fills them out as before.

ibogle commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+    bfs_prop();
+    //check for potentially false articulation points
+    while(true){
+      femv->switchActiveMultiVector(); 
```
I have cleaned up the femv->switchActiveMultiVector() calls, and added a comment to indicate what switch is occurring.

ibogle commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+  const Teuchos::ArrayRCP<const scalar_t>& propagate(void){ 
+    //run bfs_prop
+    bfs_prop();
+    //check for potentially false articulation points
```
I re-wrote all comments in this function to better reflect what was happening.

ibogle commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+
+  }
+  //propagation functions
+  //returns a flag for each vertex:
```
I have revised these comments.

ibogle commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+          lno_t* outs = g->out_vertices(curr_node.id);
+          for(int j = 0; j < out_degree; j++){
+            label_t neighbor = femvData[outs[j]];
+            if(neighbor.getGroundingStatus() == ICEPROPGS_HALF && neighbor.first_label != mapWithCopies->getGlobalElement(curr_node.id) && neighbor.first_sender == mapWithCopies->getGlobalElement(curr_node.id)){
```
Added comments.

ibogle commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+
+  //performs one bfs_prop iteration (does not check for incomplete propagation)
+  void bfs_prop(){
+    //do
```
There is old pseudocode at the beginning of this function. I'll remove it and add more descriptive comments.

ibogle commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+      }
+      
+      //store old vertex statuses for detecting changes
+      std::vector<IcePropGrounding_Status> old_statuses(nLocalOwned,ICEPROPGS_NONE);      
```
I have moved the allocation outside the loop, at the beginning of the function. I also updated the comments.

ibogle commented on packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp:
```diff
@@ -0,0 +1,499 @@
+  void giveLabels(IcePropVtxLabel<lno_t,gno_t>& curr_node, IcePropVtxLabel<lno_t,gno_t>& neighbor){
+    IcePropGrounding_Status curr_gs = curr_node.getGroundingStatus();
+    IcePropGrounding_Status nbor_gs = neighbor.getGroundingStatus();
+    int curr_node_gid = mapWithCopies->getGlobalElement(curr_node.id);
```
Done.

ibogle commented on packages/zoltan2/test/connectivity/iceSheetIntegration.cpp:
```diff
@@ -0,0 +1,218 @@
+  std::cout<<me<<": calling the utility function\n";
+  Teuchos::ArrayView<const bool> basalView = Teuchos::ArrayView<const bool>(basalFriction,nlocal);
+  Teuchos::ArrayView<const gno_t> boundaryView = Teuchos::ArrayView<const gno_t>(boundaryEdges,2*numLocalBoundaryEdges);
+  std::vector<int> status_vec(nlocal,0);
```
Specifically for the boolean instantiation of Teuchos::Array.

ibogle commented on packages/zoltan2/test/connectivity/Zoltan2_IceUtil.h:
```diff
@@ -0,0 +1,186 @@
+  //get the list of boundary edges instead of flags representing articulation points,
+  //the integration code takes a list of boundary edges.
+  int count = 0;
+  while(fin>>a>>b>>throwaway){
```
These have been removed from the format. They were a hold-over from the actual .quad.msh format that I got from Mauro. I just copied the format and re-used my file readers.

ibogle commented on packages/zoltan2/test/connectivity/Zoltan2_IceUtil.h:
```diff
@@ -0,0 +1,186 @@
+  for(int i = 0; i <= nodes; i++){
+    std::getline(fin, throwaway);
+  }
+  for(int i = 0; i < skip2; i++){
```
They have been removed.

ibogle commented on packages/zoltan2/test/connectivity/Zoltan2_IceUtil.h:
```diff
@@ -0,0 +1,186 @@
+    exit(0);
+  }
+  std::string throwaway;
+  fin>>throwaway>>throwaway;
```
I have improved the comments here generally, but I've also added this specific comment.

mperego commented:
@ibogle @kddevin any update on this? This capability will become critical in the next months.

sebrowne commented:
Closing in favor of #10677
